### ISSUE TITLE
env_mach_pes.xml is too verbose

### ISSUE THREAD

**Issue by amametjanov** [ESMCI/cime#949]
The old CIME2 PE file used to be ~126 lines long and the bottom 62 lines contained all of the relevant information that used to fit in one screen without any scrolling required for editing.

The new CIME5 PE file is now ~325 lines long with repetitive `type, desc, valid_values` XML tags that rarely need to be changed and stay the same from run to run. But to edit `NTASKS_*`, `NTHRDS_*`, and `ROOTPE_*` I have to scroll through 4 screens: not efficient and prone to errors.

I'd prefer to revert to the old format: i.e.
```
<entry id="NTASKS_ATM"   value="21600"  />    
<entry id="NTHRDS_ATM"   value="16"  />    
<entry id="ROOTPE_ATM"   value="0"  />    
<entry id="NINST_ATM"   value="1"  />    
<entry id="NINST_ATM_LAYOUT"   value="concurrent"  />    

<entry id="NTASKS_LND"   value="4096"  />    
<entry id="NTHRDS_LND"   value="16"  />    
<entry id="ROOTPE_LND"   value="16480"  />    
<entry id="NINST_LND"   value="1"  />    
<entry id="NINST_LND_LAYOUT"   value="concurrent"  />    

<entry id="NTASKS_ICE"   value="21600"  />    
<entry id="NTHRDS_ICE"   value="16"  />    
<entry id="ROOTPE_ICE"   value="0"  />    
<entry id="NINST_ICE"   value="1"  />    
<entry id="NINST_ICE_LAYOUT"   value="concurrent"  />    

<entry id="NTASKS_OCN"   value="11168"  />    
<entry id="NTHRDS_OCN"   value="12"  />    
<entry id="ROOTPE_OCN"   value="21600"  />    
<entry id="NINST_OCN"   value="1"  />    
<entry id="NINST_OCN_LAYOUT"   value="concurrent"  />    

<entry id="NTASKS_CPL"   value="21600"  />    
<entry id="NTHRDS_CPL"   value="16"  />    
<entry id="ROOTPE_CPL"   value="0"  />    

<entry id="NTASKS_GLC"   value="1"  />    
<entry id="NTHRDS_GLC"   value="1"  />    
<entry id="ROOTPE_GLC"   value="2"  />    
<entry id="NINST_GLC"   value="1"  />    
<entry id="NINST_GLC_LAYOUT"   value="concurrent"  />    

<entry id="NTASKS_ROF"   value="1024"  />    
<entry id="NTHRDS_ROF"   value="16"  />    
<entry id="ROOTPE_ROF"   value="20576"  />    
<entry id="NINST_ROF"   value="1"  />    
<entry id="NINST_ROF_LAYOUT"   value="concurrent"  />    

<entry id="NTASKS_WAV"   value="1"  />    
<entry id="NTHRDS_WAV"   value="1"  />    
<entry id="ROOTPE_WAV"   value="3"  />    
<entry id="NINST_WAV"   value="1"  />    
<entry id="NINST_WAV_LAYOUT"   value="concurrent"  />    

<entry id="PSTRID_ATM"   value="1"  />    
<entry id="PSTRID_LND"   value="1"  />    
<entry id="PSTRID_ICE"   value="1"  />    
<entry id="PSTRID_OCN"   value="1"  />    
<entry id="PSTRID_CPL"   value="1"  />    
<entry id="PSTRID_GLC"   value="1"  />    
<entry id="PSTRID_ROF"   value="1"  />    
<entry id="PSTRID_WAV"   value="1"  />    

<entry id="TOTALPES"   value="488548"  />    
<entry id="PES_LEVEL"   value="1r"  />    
<entry id="MAX_TASKS_PER_NODE"   value="64"  />    
<entry id="PES_PER_NODE"   value="$MAX_TASKS_PER_NODE"  />    
<entry id="COST_PES"   value="0"  />    
<entry id="CCSM_PCOST"   value="-8"  />    
<entry id="CCSM_TCOST"   value="0"  />    
<entry id="CCSM_ESTCOST"   value="-4"  />    
```

**Comment by jedwards4b** [comment]
Fixed in #932 already on master

**Comment by rljacob** [comment]
@amametjanov try building an X or A case with latest master, look at the env_mach_pes, and let us know.   See also issue #890

**Comment by amametjanov** [comment]
The new format in cime master is not the same as before and it's not clear why the structure is now different. Previously, components' tasks, threads and root were naturally grouped by component. Now, the grouping is by tasks, or threads or ...
```
  <group id="mach_pes">
    <entry id="NTASKS">
      <type>integer</type>
      <values>
        <value component="ATM">64</value>
        <value component="CPL">64</value>
        <value component="OCN">64</value>
        <value component="WAV">64</value>
        <value component="GLC">64</value>
        <value component="ICE">64</value>
        <value component="ROF">64</value>
        <value component="LND">64</value>
        <value component="ESP">1</value>
      </values>
      <desc>number of tasks for each component</desc>
    </entry>
    <entry id="NTHRDS">
      <type>integer</type>
      <values>
        <value component="ATM">1</value>
        <value component="CPL">1</value>
        <value component="OCN">1</value>
        <value component="WAV">1</value>
        <value component="GLC">1</value>
        <value component="ICE">1</value>
        <value component="ROF">1</value>
        <value component="LND">1</value>
        <value component="ESP">1</value>
      </values>
      <desc>number of threads for each task in each component</desc>
    </entry>
    <entry id="ROOTPE">
      <type>integer</type>
      <values>
        <value component="ATM">0</value>
        <value component="CPL">0</value>
        <value component="OCN">0</value>
        <value component="WAV">0</value>
        <value component="GLC">0</value>
        <value component="ICE">0</value>
        <value component="ROF">0</value>
        <value component="LND">0</value>
        <value component="ESP">0</value>
      </values>
      <desc>ROOTPE (mpi task in MPI_COMM_WORLD) for each component</desc>
    </entry>
```

**Comment by jedwards4b** [comment]
Why the structure is now different:
Try the following commands
 $xmlquery NTASKS 

$ xmlquery NTASKS NTHRDS ROOTPE

$ xmlquery NTASKS_ATM

$ xmlchange NTASKS=8

$ xmlchange NTASKS_ATM=16

$ xmlquery NTASKS

**Comment by rljacob** [comment]
Can we do something like this for env_mach_pes?

`  <group id="mach_pes">
    <entry id="LAYOUT">
      <type>integer</type>
      <values>
        <value component="ATM">64,1, 0</value>
       <value component="OCN">64,1, 0</value>
       (etc.)
`

That way everything is grouped together and easy to see.

**Comment by amametjanov** [comment]
This use case demonstrates being able to change tasks or threads for all components: i.e. `xmlchange NTASKS=8`. I rarely need to do that.

I think it'd be better to keep the old textual format with grouping by component. The in-memory/Python structure can be new, but the new text forces everyone to adjust.

**Comment by jedwards4b** [comment]
DId you try all of the examples I suggested?  You can still change one at a time by two different methods.   

xmlchange NTASKS_ATM=8 or xmlchange --subgroup ATM NTASKS=8

**Comment by amametjanov** [comment]
Another possible "concise" format could be
```
<values>
 <value component="ATM">64x1 @ 0</value>
 <value component="LND">64x1 @ 0</value>
 <value component="ICE">64x1 @ 0</value>
 <value component="OCN">64x1 @ 0</value>
 <value component="ROF">64x1 @ 0</value>
 <value component="CPL">64x1 @ 0</value>
 <value component="GLC">64x1 @ 0</value>
 <value component="WAV">64x1 @ 0</value>
 <value component="ESP">64x1 @ 0</value>
</values>
 ```

**Comment by mvertens** [comment]
I think that the new format will be much easier to determine layouts of the
system. I think it is the right step forwards. Its true - it will take some
adjustment.

On Wed, Dec 21, 2016 at 4:33 PM, Azamat Mametjanov <notifications@github.com

**Comment by gold2718** [comment]
This conversation seems to imply that people are hand editing xml files instead of using xmlchange. xmlchange has the advantages of being easy to script and also produces documentation of the event in the CaseStatus file.
Why is anyone still hand editing these files?

**Comment by amametjanov** [comment]
@jedwards4b, yes, tried all examples in all three versions: cime2 in acme, cime5 in acme, cime master. The only new capability is being able to change all tasks at once. Changing tasks for a single component is retained and that's good.

If we can't keep the old format, then it could be easier to use the existing `NTasks x NThreads` syntax (used in tests) and add an element for the Root PE.

**Comment by gold2718** [comment]
I think the real advantage of the new format is that it enables commands such as
./xmlquery NTASKS
I understand making layout changes in order to improve the user interface, however, xmlquery and xmlchange are our user interface, not vi.

**Comment by amametjanov** [comment]
Hand-editing is actually easier and quicker for me. Going through xmlchange either requires keeping two terminal sessions open: one to see what the previous layout was and another to actually do the xmlchange. Or a sequence of open, view, close, xmlchange, open, view, close ... 

After all edits are done, all that's needed is to share the PE file itself. Sharing the xmlchange script works too.

**Comment by gold2718** [comment]
The xmlchange script has the advantage of being portable across any xml layout changes, the same command sequence works in cime2 as cime5 (although new cime5 options are obviously not backwards compatible to cime2).
That plus the documentation in CaseStatus speaks to converting to xmlchange (IMHO).

**Comment by mvertens** [comment]
We are actually strongly encouraging scientists to never hand-edit the xml
files for several reasons. First, xmlchange does error checking and
prevents users from making mistakes that are not caught. Second, xmlchange
commands can easily be incorporated into a script that can be used to
create multiple experiments out of the box easily.

**Comment by rljacob** [comment]
What's missing in the xmlchange workflow is the view of the complete layout (tasks x threads x root for each model) that one needs to have in front of them when they are trying to load-balance the full system or experiment.  That is absolutely crucial to tell what you're doing.  In a vi window, I can see it all at once.  A series of xmlchange commands is too isolating.

But here's a suggestion:  add a --printlayout option to xmlchange which prints out the tasks, threads, root in a nice table.  A performance engineer can then sit at the prompt and issue: "xmlchange NTASKS_ATM=8  --printlayout" and get instant feedback on what the complete layout looks like after that one change.

**Comment by ndkeen** [comment]
That might work.  After recently going thru a lot of PE layout lore, I ended up using a spreadsheet.  In general, I feel like we could benefit from a graphic of some sort describing the layout.  I can imagine a .jpg as well as some condensed text.  Extra credit: find a way of merging a performance summary result with a layout.

**Comment by gold2718** [comment]
I have two issues with a -printlayout option. First, you have to call xmlchange to see your layout. Second, different folks probably want a different layout. I suggest just using a script to parse the output in a way you want. Here is a gist of a simple bash script to create a layout in a case directory:
https://gist.github.com/gold2718/2b6f806a924465ad2cfc0d30dd29db99

**Comment by rljacob** [comment]
--printlayout could also be an option to xmlquery.  There's probably lots of ways to do this.  But we need a way to easily print a summary table of the tasks, threads, etc and then reprint that when we make a change.

**Comment by jedwards4b** [comment]
```
./xmlquery NTASKS NTHRDS ROOTPE

Results in group mach_pes
	NTASKS: ['CPL:1728', 'ATM:1728', 'LND:1088', 'ICE:640', 'OCN:160', 'ROF:1088', 'GLC:1', 'WAV:1', 'ESP:1']
	NTHRDS: ['CPL:2', 'ATM:2', 'LND:2', 'ICE:2', 'OCN:2', 'ROF:2', 'GLC:2', 'WAV:2', 'ESP:1']
	ROOTPE: ['CPL:0', 'ATM:0', 'LND:0', 'ICE:1088', 'OCN:1728', 'ROF:0', 'GLC:0', 'WAV:0', 'ESP:0']
```

**Comment by gold2718** [comment]
I think we should have a separate printPElayout script, probably written in python using the same infrastructure as xmlquery. If folks want, this script could then be automatically called from other scripts with a flag.
I also think similar scripts, such as printrun (which would print all the run parameters such as STOP_OPTION and REST_N) would be useful. I suggest closing this issue and opening an issue (or issues) to define the behavior of the new script(s).

**Comment by jedwards4b** [comment]
I'm not sure if we need to add a bunch of single function scripts to the repo - the problem I see when we get down to this level of detail is that one persons improvement is anothers annoyance.   I don't know how we can possibly expect to satisfy everyone.

**Comment by gold2718** [comment]
printPElayout can be pretty flexible by taking a format argument similar to the Unix date function.

**Comment by rljacob** [comment]
I don't see a slippery slope here.   Most of the xml variables in the env files are standalone and don't need special viewers/printers.  NTASK, NTHREAD, ROOTPE are different because they need to be seen together.

**Comment by billsacks** [comment]
Discussions like this frustrate me because we experienced developers are trying to speculate about what might be most helpful to users. It feels like we keep having these same discussions of a UI nature, with no real conclusions as far as I can tell.

We are lucky enough to share a building with a bunch of users with varying experience levels. Why not recruit 5 or 10 people with varying experience levels for a discussion of this? -- both this PE-specific issue and the more general questions of how to interact with the xml files. I'd be willing to help with this if others would find it helpful.

**Comment by jedwards4b** [comment]
@billsacks I think that it's too early to carry on such a survey.  I would be agreeable to doing this only after the users have considerable experience with cesm2, if we do it too early there is too much of a tendency to wish for the "good ol days"    A survey of users 6 months after the release might be suitable.

**Comment by rljacob** [comment]
In this particular case a user, Az, who spends a LOT of time tuning pe-layouts for ACME cases on many machines, is finding the new env_mach_pes file a pain to work with.  The latest and greatest on master is an improvement (less XML tag clutter) but still not optimal.  I think we should do the printlayout as a quick fix and we can think again about the exact XML format of that file.

Wasn't the cime2 version not considered proper XML and that was another motivation for the change?

**Comment by billsacks** [comment]
> @billsacks I think that it's too early to carry on such a survey. I would be agreeable to doing this only after the users have considerable experience with cesm2, if we do it too early there is too much of a tendency to wish for the "good ol days" A survey of users 6 months after the release might be suitable.

I feel this is a backward approach to UI design - releasing the software then inviting input on the UI afterwards. But I know that we all have a lot going on, so won't press this issue further right now.

> In this particular case a user, Az, who spends a LOT of time tuning pe-layouts for ACME cases on many machines, is finding the new env_mach_pes file a pain to work with.

For the record, I certainly felt the same way prior to the changes made over the last week - though haven't played with the recent version enough to be able to weigh in on it. (It's definitely a BIG improvement over what was there before; I just don't know if it is as easy to use as hand-editing the cime2 xml file.) My frustration was not directed at Az, who I feel raised a fair point.

I'll also add that the format I've found most useful for confirming my PE layouts is the comment that used to get output to the top of the run script, which told you the PE range on which each component was running. This was very useful for confirming that your NTASKS and ROOTPE settings were in sync.

**Comment by jedwards4b** [comment]
See issue #953

**Comment by amametjanov** [comment]
All I am asking to fix in this issue is to place NTasks, NThreads and RootPE values as close together as possible: see for examples https://github.com/ESMCI/cime/issues/949#issuecomment-268669279 or https://github.com/ESMCI/cime/issues/949#issuecomment-268673594.

**Comment by gold2718** [comment]
I think this issue should remain open but only for discussion of any change to the env_mach_pes.xml format. For the record, I approve of the current format. Does anyone have a reason why a different XML format would be an improvement? The XML file is not designed for viewing or editing so I don't view those as good reasons for a change.

**Comment by rljacob** [comment]
It should be possible to have an XML format that supports the xmlchange/xmlquery tools **and** lets you see it all together in a when viewing in an editor.  We'll make some suggestions.

**Comment by rljacob** [comment]
You could still close this issue because the verbose xml tags are gone on the latest master.

**Comment by jedwards4b** [comment]
Details of the xml format should be beyond the scope of the UI.

**Comment by gold2718** [comment]
I agree which is why this is a separate issue. The XML format should be optimized based on the best database principles. Does anyone have an XML design which removes a problem with the current design?

**Comment by jedwards4b** [comment]
What should matter to the UI is the python API to the case object and in turn the XML file interface objects, not the format of the xml file itself.  This will make the tool far more extensible.  Requirements that the xml file be human readable will, I believe, unnecessarily limit our ability to improve both the xml format and the code in the interface.    There are still some variables in the env files that xmlquery and xmlchange do not see and cannot adjust - extending the capability of these interfaces to handle these other variables is to me a far more important goal.
